torchao/quantization/quant_primitives.py

```diff
@@ -1039,6 +1044,172 @@ def reshape_w(w):
     return q_w, s_group, s_channel, w_ref
 
 
+def choose_qparams_gguf(
+    input: Optional[torch.Tensor],
+    block_size: List[int],
+    target_dtype: torch.dtype,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    # there are two sets of qparams: quantized_block_scale, quantized_block_min and super_block_scale_scale and super_block_min_scale
+    # the relationship is the following:
+    # block_scale = quantized_block_scale * super_block_sclae
+    # block_min = quantized_block_min * super_block_min
+    # quantized_val = (float_val - block_min) / block_scale + quant_min
+    # first we calculate block_scale and block_min
+    # then we calculate super_block_scale_scale and super_block_min_scale
+    # after that we can calculate quantized_block_scale and quantized_min_scale
+    # the returned values are: super_block_scale_scale, super_block_min_scale, quantized_block_scale
+    # and quantized_min_scale
+
+    # 1. get block_scale block_min
+    shape_for_reduction, reduction_dims = _get_reduction_params(
+        block_size, input.size()
+    )
+    input = input.view(shape_for_reduction)
+    min_val = torch.amin(input, dim=reduction_dims, keepdim=False)
+    max_val = torch.amax(input, dim=reduction_dims, keepdim=False)
+    quant_max = 15
+    quant_min = 0
+    # asymmetric quant to fully utilize the range
+    block_scale = max_val / (float(quant_max - quant_min) / 2)
+    block_scale = (max_val - min_val) / float(quant_max - quant_min)
+    block_min = min_val
+
+    # 2. get super_block_scale_scale and super_block_min_scale
+    block_scale_clone = block_scale.clone()
+    block_min_clone = block_min.clone()
+
+    assert _GGUF_QK_K % block_size[-1] == 0
+    super_block_size = (1, _GGUF_QK_K // block_size[-1])
+    shape_for_reduction, reduction_dims = _get_reduction_params(
+        super_block_size, block_scale_clone.size()
+    )
+    block_scale_clone = block_scale_clone.view(shape_for_reduction)
+    block_min_clone = block_min_clone.view(shape_for_reduction)
+
+    shape_after_reduction = shape_for_reduction.copy()
+    for i in reduction_dims:
+        shape_after_reduction[i] = 1
+
+    block_scale_absmax = torch.amax(
+        torch.abs(block_scale_clone), dim=reduction_dims, keepdim=False
+    )
+    block_min_absmax = torch.amax(
+        torch.abs(block_min_clone), dim=reduction_dims, keepdim=False
+    )
+
+    # 2. get super_block_scale_scale and super_block_min_scale
+    quant_max = 2**6 - 1
```

Comment:
is this for 6 bit?

Reply:
oh nevermind. this is for 6bit scale value. rename this appropriately else it is confusing